lib/mongo/cluster.rb

```diff
@@ -89,14 +86,15 @@ def add(host)
     # @since 2.0.0
     def initialize(seeds, options = {})
       @addresses = []
+      @addresses_update = Mutex.new
       @servers = []
+      @servers_update = Mutex.new
```

Comment:
I'd just put one big lock around Cluster and do all non-I/O operations (both reads and updates to all data structures) while holding the lock. If you do I/O, drop the lock, do your operation without touching any Cluster data structures, then reacquire the lock in order to update your data structures according to what you discovered from doing the I/O. So, no need to distinguish between these two mutexes, just lock the whole Cluster whenever you touch it.